Use Conan for dependencies

ignotuscaligo opened the pull request:
### What

* Use Conan for managing external dependencies
  - Remove submodules, make `libpng` linking less brittle.
  - Update README.
* Use updated API from `tinyobjloader`
  - `warn` parameter was dropped from `::LoadObj`, warnings are now included in the error string.
* Set `-march=native` compile option for `ray-tracer`
* Use `Vector` union to access components of `__m256d`
  - `m256d_f64` array is not portable.
* Update compiling and linking options for Unix and gcc
  - `std::thread` requires additional configuration to work correctly on
Unix. CMake provides a package `Threads` that sets up the appropriate
libraries.
  - `std::filesystem` requires an additional library for gcc, `stdc++fs`.
It seems like `std::filesystem` hasn't been combined into the main
`stdc++` library for gcc yet.
* Add missing define required for tiny obj loader

ignotuscaligo committed:
Use Conan for managing external dependencies

- Remove submodules, make `libpng` linking less brittle.
- Update README.

ignotuscaligo committed:
Use updated API from `tinyobjloader`

- `warn` parameter was dropped from `::LoadObj`, warnings are now included in the error string.

ignotuscaligo committed:
Switch to `find_package` and virtual environment configuration for Conan

- Better matches normal expected interaction with libraries, previous configuration would be confusing when trying to read documentation for the external libraries.
- Add some comments separating each section of the cmake file.
- Explicitly set `CMAKE_MODULE_PATH` so that cmake knows where to look for the `FindX` files.
- Update README.

cdelguercio commented:
```
/usr/bin/ld: CMakeFiles/ray-tracer.dir/src/ObjReader.cpp.o: in function `ObjReader::loadMeshes(std::filesystem::__cxx11::path const&)':
ObjReader.cpp:(.text+0x167): undefined reference to `tinyobj::LoadObj(tinyobj::attrib_t*, std::vector<tinyobj::shape_t, std::allocator<tinyobj::shape_t> >*, std::vector<tinyobj::material_t, std::allocator<tinyobj::material_t> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, char const*, char const*, bool)'
```

ignotuscaligo committed:
Set `-march=native` compile option for `ray-tracer`

ignotuscaligo committed:
Use `Vector` union to access components of `__m256d`

- `m256d_f64` array is not portable.

ignotuscaligo committed:
Add missing define required for tiny obj loader

ignotuscaligo committed:
Update compiling and linking options for Unix and gcc

- `std::thread` requires additional configuration to work correctly on
Unix. CMake provides a package `Threads` that sets up the appropriate
libraries.
- `std::filesystem` requires an additional library for gcc, `stdc++fs`.
It seems like `std::filesystem` hasn't been combined into the main
`stdc++` library for gcc yet.

ignotuscaligo commented:
> ```
> /usr/bin/ld: CMakeFiles/ray-tracer.dir/src/ObjReader.cpp.o: in function `ObjReader::loadMeshes(std::filesystem::__cxx11::path const&)':
> ObjReader.cpp:(.text+0x167): undefined reference to `tinyobj::LoadObj(tinyobj::attrib_t*, std::vector<tinyobj::shape_t, std::allocator<tinyobj::shape_t> >*, std::vector<tinyobj::material_t, std::allocator<tinyobj::material_t> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, char const*, char const*, bool)'
> ```

This was fixed by adding `#define TINYOBJLOADER_IMPLEMENTATION` before `#include <tiny_obj_loader.h>`, apparently required to set up the types correctly.

cdelguercio reviewed (approved):
LGTM will check later
